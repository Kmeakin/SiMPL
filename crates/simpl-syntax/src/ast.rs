pub type Ident = String;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        val: Lit,
    },
    Var {
        name: String,
    },
    If {
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        bindings: Vec<(Ident, Self)>,
        body: Box<Self>,
    },
    Letrec {
        bindings: Vec<(Ident, Self)>,
        body: Box<Self>,
    },
    Lambda {
        params: Vec<Ident>,
        body: Box<Self>,
    },
    App {
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Bool(bool),
    Int(i64),
    Float(f64),
}
