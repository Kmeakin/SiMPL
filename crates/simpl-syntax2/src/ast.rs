pub type Ident = String;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        val: Lit,
    },
    Var {
        name: String,
    },
    If {
        test: Box<Expr>,
        then_branch: Box<Expr>,
        else_branch: Box<Expr>,
    },
    Let {
        bindings: Vec<(Ident, Expr)>,
        body: Box<Expr>,
    },
    Lambda {
        params: Vec<Ident>,
        body: Box<Expr>,
    },
    App {
        func: Box<Expr>,
        arg: Box<Expr>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Bool(bool),
    Int(i64),
    Float(f64),
}
