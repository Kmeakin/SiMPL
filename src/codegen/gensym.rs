use crate::util::counter::{Counter, FromId};

#[derive(Debug, Copy, Clone)]
pub struct Gensym {
    counter: Counter<u32>,
    prefix: &'static str,
}

impl Gensym {
    pub fn new(prefix: &'static str) -> Self {
        Self {
            counter: Counter::new(),
            prefix,
        }
    }

    pub fn current(&self) -> String {
        format!("{}{}", self.prefix, self.current())
    }

    pub fn next(&mut self) -> String {
        let x = self.next();
        format!("{}{}", self.prefix, x)
    }
}
