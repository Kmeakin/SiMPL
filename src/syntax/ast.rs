pub use crate::types::ty::Type;
use derive_more::Display;
pub use simple_symbol::Symbol;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        val: Lit,
    },
    Var {
        name: Symbol,
    },
    Binop {
        lhs: Box<Self>,
        rhs: Box<Self>,
        op: Binop,
    },
    If {
        test: Box<Self>,
        then: Box<Self>,
        els: Box<Self>,
    },
    Let {
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Letrec {
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        params: Vec<Param>,
        body: Box<Self>,
    },
    App {
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq, Display)]
pub enum Lit {
    #[display(fmt = "{}", _0)]
    Bool(bool),
    #[display(fmt = "{}", _0)]
    Int(i64),
    #[display(fmt = "{}", _0)]
    Float(f64),
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub name: Symbol,
    pub ann: Option<Type>,
    pub val: Box<Expr>,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Binop {
    IntAdd,
    IntSub,
    IntMul,
    IntDiv,
    IntLt,
    IntLeq,
    IntGt,
    IntGeq,
    FloatAdd,
    FloatSub,
    FloatMul,
    FloatDiv,
    FloatLt,
    FloatLeq,
    FloatGt,
    FloatGeq,
    Eq,
    Neq,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub name: Symbol,
    pub ann: Option<Type>,
}

impl From<bool> for Lit {
    fn from(other: bool) -> Self {
        Self::Bool(other)
    }
}

impl From<i64> for Lit {
    fn from(other: i64) -> Self {
        Self::Int(other)
    }
}

impl From<f64> for Lit {
    fn from(other: f64) -> Self {
        Self::Float(other)
    }
}
