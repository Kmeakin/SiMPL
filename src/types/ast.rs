pub use crate::syntax::ast::{Ident, Lit};
use crate::{
    syntax::ast::Expr,
    types::ty::{Type, TypeEnv},
    util::counter::{Counter, FromId},
};
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Ident,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Ident,
    pub val: Box<TypedExpr>,
}

pub type TypeVarGen = Counter<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Self {
        Self::Var(id)
    }
}

impl FromStr for TypedExpr {
    type Err = String;
    fn from_str(src: &str) -> Result<Self, Self::Err> {
        // TODO: return a trait object instead of unwrapping
        let ast = crate::syntax::parse(src).unwrap();
        Self::from_ast(ast)
    }
}

impl TypedExpr {
    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => {
                let (bindings, body) = expand_let(&bindings, *body);

                assert!(bindings.len() == 1);
                let ty = gen.next();

                let (binding_name, binding_val) = &bindings[0];
                let binding_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(binding_name.into(), binding_ty.clone());

                Self::Let {
                    ty,
                    binding: LetBinding {
                        ty: binding_ty,
                        name: binding_name.clone(),
                        val: box Self::from_ast_inner(binding_val.clone(), tenv, gen)?,
                    },
                    body: box Self::from_ast_inner(body, &extended_tenv, gen)?,
                }
            }
            Expr::Letrec { bindings, body } => {
                let ty = gen.next();
                let mut extended_tenv = tenv.clone();

                let mut tvars = vec![];
                for (name, _) in &bindings {
                    let tvar = gen.next();
                    extended_tenv.insert(name.clone(), tvar.clone());
                    tvars.push(tvar);
                }

                let mut new_bindings = vec![];

                for ((name, val), tvar) in bindings.into_iter().zip(tvars) {
                    new_bindings.push(LetBinding {
                        name: name.clone(),
                        ty: tvar,
                        val: box Self::from_ast_inner(val.clone(), &extended_tenv, gen)?,
                    });
                }

                Self::Letrec {
                    ty,
                    bindings: new_bindings,
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Lambda { params, body } => {
                let (params, body) = expand_lambda(&params, *body);

                assert!(params.len() == 1);
                let ty = gen.next();

                let param_name = &params[0];
                let param_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(param_name.clone(), param_ty.clone());
                Self::Lambda {
                    ty,
                    param: Param {
                        ty: param_ty,
                        name: param_name.clone(),
                    },
                    body: box Self::from_ast_inner(body, &extended_tenv, gen)?,
                }
            }
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}

/// Expand an `ast::Expr::Let` with many bound variables into a nested
/// sequences of lets each binding a single variable
/// eg `let x = 1, y = 2 in add x y` expands to
///    `let x = 1 in let y = 2 in add x y`
fn expand_let(bindings: &[(Ident, Expr)], body: Expr) -> (Vec<(Ident, Expr)>, Expr) {
    assert!(!bindings.is_empty());
    if bindings.len() == 1 {
        (bindings.to_vec(), body)
    } else {
        let binding = &bindings[0];

        let (rest_bindings, rest_body) = expand_let(&bindings[1..], body);

        (
            vec![binding.clone()],
            Expr::Let {
                bindings: rest_bindings,
                body: box rest_body,
            },
        )
    }
}

/// Expand an `ast::Expr::Lambda` with many parameters into a nested
/// sequences of lambdas each binding a single parameter
/// eg `\x, y -> add x y` expands to
///    `\x -> \y -> add x y`
fn expand_lambda(params: &[Ident], body: Expr) -> (Vec<Ident>, Expr) {
    assert!(!params.is_empty());
    if params.len() == 1 {
        (params.to_vec(), body)
    } else {
        let param = &params[0];

        let (rest_params, rest_body) = expand_lambda(&params[1..], body);

        (
            vec![param.clone()],
            Expr::Lambda {
                params: rest_params,
                body: box rest_body,
            },
        )
    }
}

#[test]
fn test_expand_let() {
    let bindings = vec![
        ("x".into(), Expr::Lit { val: Lit::Int(1) }),
        ("y".into(), Expr::Lit { val: Lit::Int(2) }),
    ];

    let body = Expr::Lit { val: Lit::Int(0) };

    assert_eq!(
        expand_let(&bindings, body),
        (
            vec![("x".into(), Expr::Lit { val: Lit::Int(1) }),],
            Expr::Let {
                bindings: vec![("y".into(), Expr::Lit { val: Lit::Int(2) })],
                body: box Expr::Lit { val: Lit::Int(0) }
            }
        ),
    )
}

#[test]
fn test_expand_lambda() {
    let params = vec!["x".into(), "y".into()];
    let body = Expr::Lit { val: Lit::Int(0) };

    assert_eq!(
        expand_lambda(&params, body),
        (
            vec!["x".into()],
            Expr::Lambda {
                params: vec!["y".into()],
                body: box Expr::Lit { val: Lit::Int(0) }
            }
        ),
    )
}
