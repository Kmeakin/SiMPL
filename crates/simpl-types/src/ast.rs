use crate::{
    ty::{Type, TypeEnv},
    FromId, IdGen,
};
use simpl_syntax::ast::Expr;
pub use simpl_syntax::ast::{Ident, Lit};

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Ident,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Ident,
    pub val: Box<TypedExpr>,
}

pub type TypeVarGen = IdGen<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Type {
        Type::Var(id)
    }
}

impl TypedExpr {
    pub fn from_str(src: &str) -> Result<Self, String> {
        // TODO: return a trait object instead of unwrapping
        let ast = simpl_syntax::parse(src).unwrap();
        Self::from_ast(ast)
    }

    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => {
                assert!(bindings.len() >= 1);
                let ty = gen.next();

                // TODO: desugar `let v1 = e1, v2 = e2 in b` into
                // `let v1 = e1 in let v2 = e2 in b`
                let (binding_name, binding_val) = &bindings[0];
                let binding_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(binding_name.into(), binding_ty.clone());

                Self::Let {
                    ty,
                    binding: LetBinding {
                        ty: binding_ty,
                        name: binding_name.clone(),
                        val: box Self::from_ast_inner(binding_val.clone(), tenv, gen)?,
                    },
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Letrec { bindings, body } => {
                let ty = gen.next();
                let mut extended_tenv = tenv.clone();

                let mut tvars = vec![];
                for (name, _) in &bindings {
                    let tvar = gen.next();
                    extended_tenv.insert(name.clone(), tvar.clone());
                    tvars.push(tvar);
                }

                let mut new_bindings = vec![];

                for ((name, val), tvar) in bindings.into_iter().zip(tvars) {
                    new_bindings.push(LetBinding {
                        name: name.clone(),
                        ty: tvar,
                        val: box Self::from_ast_inner(val.clone(), &extended_tenv, gen)?,
                    });
                }

                Self::Letrec {
                    ty,
                    bindings: new_bindings,
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Lambda { params, body } => {
                assert!(params.len() >= 1);
                let ty = gen.next();

                // TODO: desugar `\x, y -> e` into `\x -> \y -> e`
                let param_name = &params[0];
                let param_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(param_name.clone(), param_ty.clone());
                Self::Lambda {
                    ty,
                    param: Param {
                        ty: param_ty,
                        name: param_name.clone(),
                    },
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}
