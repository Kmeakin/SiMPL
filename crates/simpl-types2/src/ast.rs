use crate::{
    ty::{Type, TypeEnv},
    FromId, IdGen,
};
use simpl_syntax2::ast::Expr;
pub use simpl_syntax2::ast::{Ident, Lit};

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: (Type, Ident, Box<Self>),
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<(Type, Ident, Self)>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: (Type, Ident),
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

pub type TypeVarGen = IdGen<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Type {
        Type::Var(id)
    }
}

impl TypedExpr {
    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => todo!(),
            Expr::Letrec { bindings, body } => todo!(),
            Expr::Lambda { params, body } => todo!(),
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}
