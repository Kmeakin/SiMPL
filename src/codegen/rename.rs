//! `TypedExpr` -> `TypedExpr` pass
//! Alpha-renames exprs, so that no variable masks another in its enclosing
//! scope

use super::gensym::Gensym;
use crate::types::ast::TypedExpr;
pub use crate::types::ast::{Ident, LetBinding, Lit};
use lazy_static::lazy_static;
use std::{collections::HashSet, sync::Mutex};

lazy_static! {
    static ref GENSYM: Mutex<Gensym> = Mutex::new(Gensym::new("$"));
}

type Expr = TypedExpr;

fn hset<T: Eq + std::hash::Hash>(x: T) -> HashSet<T> {
    let mut hs = HashSet::new();
    hs.insert(x);
    hs
}

fn free_vars(expr: &Expr) -> HashSet<Ident> {
    match expr {
        Expr::Lit { .. } => HashSet::new(),
        Expr::Var { name, .. } => hset(name.into()),
        Expr::If {
            test,
            then_branch,
            else_branch,
            ..
        } => &(&free_vars(test) | &free_vars(then_branch)) | &free_vars(else_branch),
        Expr::Let { binding, body, .. } => {
            &free_vars(&*binding.val) | &(&free_vars(body) - &hset(binding.name.clone()))
        }
        Expr::Letrec { bindings, body, .. } => {
            &bindings
                .iter()
                .fold(free_vars(body), |acc, b| &acc | &free_vars(&*b.val))
                - &bindings
                    .iter()
                    .map(|b| b.name.clone())
                    .collect::<HashSet<_>>()
        }
        Expr::Lambda { param, body, .. } => &free_vars(body) - &hset(param.name.clone()),
        Expr::App { func, arg, .. } => &free_vars(func) | &free_vars(arg),
    }
}

impl Expr {
    fn is_alpha_eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Lit { val: x, .. }, Self::Lit { val: y, .. }) => x == y,
            (Self::Var { name: x, .. }, Self::Var { name: y, .. }) => x == y,
            (
                Self::If {
                    test: test1,
                    then_branch: then1,
                    else_branch: else1,
                    ..
                },
                Self::If {
                    test: test2,
                    then_branch: then2,
                    else_branch: else2,
                    ..
                },
            ) => test1.is_alpha_eq(test2) && then1.is_alpha_eq(then2) && else1.is_alpha_eq(else2),
            (
                Self::Let {
                    binding: binding1,
                    body: body1,
                    ..
                },
                Self::Let {
                    binding: binding2,
                    body: body2,
                    ..
                },
            ) => body1.is_alpha_eq(body2),
            (Self::Letrec { .. }, Self::Letrec { .. }) => todo!(),
            (Self::Lambda { .. }, Self::Lambda { .. }) => todo!(),
            (Self::App { .. }, Self::App { .. }) => todo!(),
            _ => false,
        }
    }
}

mod test {
    use super::*;
    use maplit::hashset as hset;
    use std::str::FromStr;

    #[track_caller]
    fn test_free_vars(src: &str, expected: HashSet<String>) {
        let ast = Expr::from_str(src).unwrap();
        let free = free_vars(&ast);
        assert_eq!(free, expected);
    }

    #[test]
    fn free_vars_lit() {
        test_free_vars("1", hset![]);
    }

    #[test]
    fn free_vars_if() {
        test_free_vars(
            "if abc then def else ghi",
            hset!["abc".into(), "def".into(), "ghi".into()],
        );
    }

    #[test]
    fn free_vars_let() {
        test_free_vars("let x = 5 in add x y", hset!["y".into()]);
    }

    #[test]
    fn free_vars_letrec() {
        todo!()
    }

    #[test]
    fn free_vars_lambda() {
        test_free_vars(r"\x -> x", hset![]);
        test_free_vars(r"\x -> y", hset!["y".into()]);
    }

    #[test]
    fn free_vars_app() {
        test_free_vars("f x", hset!["x".into(), "y".into()]);
    }
}
