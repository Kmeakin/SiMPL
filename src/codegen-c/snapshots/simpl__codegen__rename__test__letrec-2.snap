---
source: src/codegen/rename.rs
expression: renamed.pretty()
---
letrec f$0 = \x$2 -> (f$0 x$2),
       g$1 = \y$3 -> (g$1 y$3)
in
    (g$1 f$0)
